x86: rename Mellanox Spectrum network interfaces

tk154 opened the pull request:
This PR focuses on correctly labeling the `Mellanox Spectrum SN2100`, `SN2700`, `SN3420`, and `SN3700` Switches. Currently, the interfaces of those switches are all labeled `eth*`. Their order doesn't match the faceplate and is different for each model.

The management port(s) are renamed during boot based on their PCI address.
The boot flag is set for the `e1000e` driver to make this possible for the `SN2700` management ports.

However, this is not enough for the QSFP ports because they support port splitting via breakout cables during runtime. After such ports are split, their port naming begins again with `eth*`, and the same is true after they are unsplit again.
A hotplug script is used here that reads from the sysfs file `phys_port_name`, which contains `p1`, `p2`, `p3`, ... for unsplit ports and `p1s0`, `p1s1`, `p1s2` for split ports.

For the default network config, the management and QSFP ports are put into the `br-lan` bridge.
I added a uci function `ucidef_set_interface_netdev_range` to add all QSFP ports to the bridge more comfortably.

tk154 committed:
kernel: modules: load e1000e driver during pre-init stage

Set the pre-init boot flag for the e1000e network device
driver so that it is loaded during the early pre-init
stage of the boot process instead of later.

Then, it is possible to rename its network ports via the
ucidef_set_network_device_path function from the
02_network script while booting.

Signed-off-by: Til Kaiser <mail@tk154.de>
Link: https://github.com/openwrt/openwrt/pull/17251
Signed-off-by: Robert Marko <robimarko@gmail.com>

tk154 committed:
x86: rename and configure Mellanox Spectrum interfaces

This commit renames all management network ports of the
SN2100, SN2700, SN3420, and SN3700 switches based
on their PCI address during boot.

For the default network config, the management and
QSFP port(s) are put into the br-lan bridge.

Signed-off-by: Til Kaiser <mail@tk154.de>
Link: https://github.com/openwrt/openwrt/pull/17251
Signed-off-by: Robert Marko <robimarko@gmail.com>

tk154 committed:
base-files: add ucidef_set_interface_netdev_range function

The ucidef_set_interface* functions can be used to
add network interfaces to a default network configuration.
Such network interfaces often have the same base interface
name (e.g., eth* or lan*). On devices with many network ports,
adding all ports to the default config can become inconvenient.

This commit adds a new uci function ucidef_set_interface_netdev_range,
which adds network interfaces for a specific port range to a given
OpenWrt interface. The first parameter is the OpenWrt interface, the
second is the base interface name, the third is the port start, and
the fourth is the port end range.

Signed-off-by: Til Kaiser <mail@tk154.de>
Link: https://github.com/openwrt/openwrt/pull/17251
Signed-off-by: Robert Marko <robimarko@gmail.com>

blogic commented on package/base-files/files/lib/functions/uci-defaults.sh:
```diff
@@ -771,6 +771,17 @@ ucidef_add_wlan() {
 	ucidef_wlan_idx="$((ucidef_wlan_idx + 1))"
 }
 
+network_device_port_range() {
```
I think it would be better to add a function like
-> ucidef_set_interface_netdev_range wan eth 1 32
this will avoid invoking a sub shell in the last patch of this series

tk154 replied:
I just added the `ucidef_set_interface_netdev_range` function. I also removed the subshell for the `seq` call inside the function.

tk154 committed:
kernel: modules: add Mellanox Spectrum hotplug renaming script

Currently, the interfaces of the Mellanox Spectrum Switch QSFP ports
are all labeled eth*. Their order doesn't match the faceplate and is
different for each model.

They could be named during boot, but this isn't sufficient because they
support port splitting. After such ports are split, their port naming
begins again with eth*, and the same is true after they are unsplit again.

A hotplug script is used here that reads from the sysfs file
phys_port_name, which contains p1, p2, p3, ... for unsplit ports
and p1s0, p1s1, p1s2 for split ports.

Signed-off-by: Til Kaiser <mail@tk154.de>
Link: https://github.com/openwrt/openwrt/pull/17251
Signed-off-by: Robert Marko <robimarko@gmail.com>

tk154 commented:
I just removed the hotplug script from the Generic `x86/64` image and moved it to the `mlxsw-spectrum` kernel module/driver.

blogic commented:
@tk154 was on my todo to merge tonight, will post-pone till tomorrow, kindly let me know if anything else is pending

tk154 commented:
Just noticed that I forgot to check the board name inside the hotplug script.
I re-added the check with my last push; the PR is ready now.

systemcrash commented on package/kernel/linux/files/hotplug-mlxsw-spectrum-port-names.sh:
```diff
@@ -0,0 +1,10 @@
+if [ "$ACTION" = "add" ]; then
+	read -r board_name < "/tmp/sysinfo/board_name"
+
+	if [[ "$board_name" = "mellanox-technologies-ltd-msn*" ]]; then
```
```sh
	if [ "$board_name" = "mellanox-technologies-ltd-msn*" ]; then
```

systemcrash replied:
That's weird. It almost seems bash like and not sh like.

tk154 replied:
I executed the commands inside an `ash` shell. I don't have `bash` or any other shell installed on this device.
```
root@SN2100:~# echo $0
-ash
```

jonasjelonek commented:
is there anything that prevents this from being merged? (apart from doing a rebase on current main to check it still applies cleanly)

robimarko commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1026,7 +1026,7 @@ define KernelPackage/e1000e
   KCONFIG:=CONFIG_E1000E
   FILES:=$(LINUX_DIR)/drivers/net/ethernet/intel/e1000e/e1000e.ko
-  AUTOLOAD:=$(call AutoProbe,e1000e)
+  AUTOLOAD:=$(call AutoLoad,35,e1000e,1)
```
AutoProbe works with pre-init flag as well, lets not use AutoLoad if really not required

jonasjelonek commented:
@robimarko sorry for the noise. I think this is ready now :)

robimarko reviewed (commented):
Please reword the first commit to indicate that this sets the pre-init boot flag, meaning that the driver gets loaded in the pre-init stage instead of later.

The rest looks good to me

robimarko commented:
Thanks! Rebased on top of main and merged!
